qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java

```diff
@@ -332,84 +343,67 @@ public QuteTextDocument getDocument(String uri) {
-	 *                           computed value
-	 * @return the properties model for a given uri in a future and then apply the
-	 *         given function.
+	 * Version of {@link TemplateFileTextDocumentService#getTemplate} that returns a
+	 * future of the value instead of the value itself
+	 *
+	 * @param <R>                The type of the result computed by the bifunction
+	 * @param documentIdentifier the document identifier
+	 * @param code               a bifunction that accepts the parsed
+	 *                           {@link Template} and
+	 *                           a {@link CancelChecker} and returns the value to be
+	 *                           computed as a future
+	 * @see {@link TemplateFileTextDocumentService#getTemplate}
+	 * @return the result of the passed bifunction as a future
 	 */
-	public <R> CompletableFuture<R> getTemplate(QuteTextDocument document,
-			BiFunction<CancelChecker, Template, R> code) {
-		return computeModelAsync(document.getModel(), code);
-	}
-
-	private static <R, M> CompletableFuture<R> computeModelAsync(CompletableFuture<M> loadModel,
-			BiFunction<CancelChecker, M, R> code) {
-		CompletableFuture<CancelChecker> start = new CompletableFuture<>();
-		CompletableFuture<R> result = start.thenCombineAsync(loadModel, code);
-		CancelChecker cancelIndicator = () -> {
-			if (result.isCancelled())
-				throw new CancellationException();
-		};
-		start.complete(cancelIndicator);
-		return result;
-	}
-
-	private static <R, M> CompletableFuture<R> computeModelAsync2(CompletableFuture<M> loadModel,
-			BiFunction<CancelChecker, M, CompletableFuture<R>> code) {
-		CompletableFuture<CancelChecker> start = new CompletableFuture<>();
-		CompletableFuture<R> result = start.thenCombineAsync(loadModel, code)//
-				.thenCompose(a -> {
-					return a;
-				});
-		CancelChecker cancelIndicator = () -> {
-			if (result.isCancelled())
-				throw new CancellationException();
-		};
-		start.complete(cancelIndicator);
-		return result;
+	public <R> CompletableFuture<R> getTemplateCompose(TextDocumentIdentifier documentIdentifier,
+			BiFunction<Template, CancelChecker, CompletableFuture<R>> code) {
+		return documents.computeModelAsync(documentIdentifier, (template, cancelChecker) -> {
+			return Tuple.two(template, cancelChecker);
+		}).thenCompose((tuple) -> {
+			return code.apply(tuple.getFirst(), tuple.getSecond());
+		});
 	}
 
 	public void validationSettingsChanged() {
 		// trigger validation for all opened Qute template files
 		documents.all().stream().forEach(document -> {
-			triggerValidationFor((QuteTextDocument) document);
+			triggerValidationFor((QuteTextDocument) document, true);
 		});
 	}
 
 	public void dataModelChanged(JavaDataModelChangeEvent event) {
 		// trigger validation for all opened Qute template files
 		documents.all().stream().forEach(document -> {
-			triggerValidationFor((QuteTextDocument) document);
+			triggerValidationFor((QuteTextDocument) document, true);

```

Comment:
No need to delay the validation after a data model changed.

`triggerValidationFor((QuteTextDocument) document, false);`